packages/backend/src/index.ts

```diff
@@ -49,9 +49,9 @@ import { PluginEnvironment } from './types';
 function makeCreateEnv(loadedConfigs: AppConfig[]) {
   const config = ConfigReader.fromConfigs(loadedConfigs);
 
-  return (plugin: string): PluginEnvironment => {
+  return async (plugin: string): Promise<PluginEnvironment> => {
     const logger = getRootLogger().child({ type: 'plugin', plugin });
-    const database = createDatabaseClient(
+    const database = await createDatabaseClient(
```

Comment:
Tbh I think we should invert the relationship with the plugin here a bit. Instead of passing a database for every plugin, we could instead provide an interface or plain function for the plugin to ask for a database instead. That we can create databases on-demand, and we also don't need to make the code here async.

Reply:
I like this idea but I have a couple of questions. Can the plugin ask for more than one database (don't know a reason for doing that right now)? Should the plugin decide the name of the database? If not, is there a clear "identity" for a plugin that can be used?

Reply:
@Rugvip were you thinking of the discoveryApi for this?

Reply:
On that note, I also noticed the discoveryApi doesn't support async/promises either (at least not in the frontend version)

Reply:
@mfrinnstrom I'm thinking that it'd be a very plain `getMyPluginDatabase()`-type thing, so no, no multiple databases or deciding the name of the database. We'd always use the plugin ID to name the database. We've settled on backend plugin IDs being globally unique things in other places as well, like the discovery api. It doesn't prevent us from adding these things as options later, so that you could request multiple databases, but they would always be scoped to the plugin.

@andrewthauer Both the frontend and backend discovery APIs are async? ^_> The discovery APIs are very much tailored for plugins only at the moment though, I think we want to stick to other things being completely config-driven.

Reply:
@Rugvip - Sorry, I meant in the `createApiFactory.factory`.

Reply:
@andrewthauer Oh yeah those are super sync and will stay sync on the frontend :p

On the backend we've already set a precedent with the plugin `createRouter` function being async, so I don't really have a problem with that creeping that to createEnv as well.

Reply:
This sounds interesting. I was trying to keep the existing behaviour as much as possible before I rip the codebase's guts out :joy:

Where should the interface be defined/provided? Is there a "plugin registry" type module I should look into, that each plugin has access to? I don't have enough context for the discoveryApi bit here, unfortunately.

Reply:
discoveryApi would be unrelated so that's np. I think throwing this in `backend-common` should do the trick:

```ts
export type PluginDatabaseProvider = {
  getDatabase(): Promise<Knex>
}
```

You can look at `packages/backend-common/src/discovery` for inspiration, it's pretty lean and also passed through the plugin env.

Reply:
This makes sense to me. However, under the covers it might be good to still follow the suggested pattern I outlined here https://github.com/spotify/backstage/pull/2674#discussion_r497410413. That way there is a lower level model and the provider just composes that code to do it's thing :)

Reply:
I think to parallel the PluginEndpointDiscovery type, the PluginDatabaseProvider should be `getDatabase(pluginID: string): Promise<Knex>`, since we're just allowing plugins to define and declare their own identities.

This does make the abstraction slightly leaky - I can masquerade as another plugin - but I don't think the security aspect is a primary concern here, though it does remove the safeguards against database integration.